### ISSUE TITLE
[Refactor/Tests] verify components in SSR

### ISSUE THREAD

**Issue by maxpatiiuk** [Esri/calcite-design-system#10237]
### Priority impact

p - medium

### Test type

SSR

### Which Component(s)

all

### Unstable Tests

[not actionable before the migration]

When Stencil component is rendered in the Hydrate build, Stencil was providing mocks for a lot of dom APIs. ([source code](https://github.com/ionic-team/stencil/tree/8f0a8823facade0be6eabf8569831f456497c2a0/src/mock-doc))

When Lit elements are rendered in SSR, Lit also provides a mock dom API, but it's much more lightweight ([source code 1](https://github.com/lit/lit/blob/main/packages/labs/ssr-dom-shim/README.md), [source code 2](https://github.com/lit/lit/blob/main/packages/labs/ssr/src/lib/dom-shim.ts))

That means that if before Stencil would mock the `element.closest()` calls in SSR, these are no longer mocked in Lit. This may cause some Calcite components to start failing in SSR after migration (with errors like `element.closest is not a function`)

Options:
- After the migration, manually test each Calcite component in SSR
- After the migration, setup automated SSR test for each component
- Look for an alternative DOM API shim (or maybe continue using Stencil's shim if that is possible) so that most DOM APIs are still covered. Cons: complexity and slower SSR
- If you really need some DOM APIs in Node.js, the Lit team would likely be open to contributions

When you identify a place where a DOM API that is not available in SSR is accessed, you have a few options:

- Use optional chaining for property/methods:

  ```diff
  - const parent = element.closest("calcite-tree");
  + const parent = element.closest?.("calcite-tree");
  ```

- Use `globalThis` with optional chaining for global APIs:

  ```diff
  - const userAgent = navigator.userAgent;
  + const userAgent = globalThis.navigator?.userAgent;
  ```

- Conditionally execute code in browser environment only:

  ```ts
  import {isServer} from 'lit';
  
  if (isServer) {
    // only runs in server environments like Node
  } else {
    // runs in the browser
  }
  ```

  - I know that Calcite had issues in the past with the `isBrowser` check that Stencil provided. The Stencil check was based on presence/absence of DOM globals like `window.navigator` and etc. Lit's check instead is determined [at build-time](https://github.com/lit/lit/blob/a2cd76cfdea4ed717362bb1db32710d70550469d/packages/lit-html/src/is-server.ts#L16). That works thanks to a feature called [export conditions](https://nodejs.org/api/packages.html#conditional-exports) (this let's Lit set that variable to true in the node.js build, and false in the browser build)

- Tell consumers of Calcite components in SSR that they need to have happy-dom or jsdom environment setup to correctly render Calcite components.

Lit docs:
- [Authoring components with SSR in mind](https://lit.dev/docs/ssr/authoring/)
- [DOM emulation in SSR](https://lit.dev/docs/ssr/dom-emulation/)

**Comment by jcfranco** [comment]
Blocked by #10310.

**Comment by minhiu** [comment]
We also encountered issues with Lit components being used in Web Workers.

**Our use case:** We need to create web workers to generate SVGs of some infographics (that use calcite components) in the background. We import this "Infographic Builder" module (that has codes involved with Calcite components)  inside our worker, then call the function of "Infographic Builder" so that it builds the infographics in Calcite, then export to SVG. After we get the SVG (which was performed by the worker), the worker then post this result to our main thread.

**Issue**: As soon as we import this "Infographic Builder" module inside our worker, we got the issue in `reactive-element.js`

![Image](https://github.com/user-attachments/assets/6fa602b0-d96c-4a56-b818-cf13f4ed40fc)

**Steps taken**:
- We tried adding our own definition of `HTMLElement` and `document` inside our workers, but new issues keep coming up from "lit-html" and "directive-helper.js"
![Image](https://github.com/user-attachments/assets/a2c45817-7088-4b54-99c1-ee85ca90b3d5)

- We tried turning `NODE_MODE = true`, but issues will come after issues
![Image](https://github.com/user-attachments/assets/e9b4f735-cbdd-47d0-a1c0-8dcf41a5dd33)

- We also tried building the app in Production to force using `lit-html` in production mode, but no luck.
- Also tried to reconfigure WebPack to build for node, but still no luck.


cc: @jcfranco @benelan

**Comment by benelan** [comment]
Thanks for the feedback. Would it be feasible to create a simplified reproduction sample for your issue? This is the first web worker use case I've seen, and it would be nice to have something to test potential fixes.

**Comment by dingyuan** [comment]
Another note is we are using multi-thread for the work for efficiency. Can we get on a call/meeting?

**Comment by minhiu** [comment]
@benelan @jcfranco 

I have a simplified reproduction sample here: https://stackblitz.com/edit/lit-html-ssr?file=src%2Fworker.ts

- This sample simply creates a worker, then let it render a ReactiveElement.
- Please notice the console error, which is exactly what we noticed in our app.
- I have highlighted all the important parts in the screenshot:

![Image](https://github.com/user-attachments/assets/37b9bba6-4346-4d9f-9ca5-346fda83c858)

**Comment by dingyuan** [comment]
@benelan @geospatialem 

This issue is a breaking change for us. Infographic report player won't work on the server. Is it possible to fix the issue in [2025-02-06 - 3.0.0 Feb Breaking Change Release](https://github.com/Esri/calcite-design-system/milestone/163) instead of [2025-02-25 - Feb Milestone](https://github.com/Esri/calcite-design-system/milestone/164)? Otherwise we will need to go to Calcite 2 on the server, which won't include a lot of fixes we worked on for Calcite 3 version. 

Can you please check again and provide some advise? Thanks!

**Comment by jcfranco** [comment]
@dingyuan SSR has not been an officially supported feature in Calcite. It is not mentioned in the documentation or included in our examples. Based on that, we are not considering this a breaking change and won't be prioritizing it for today's 3.0 release.

**Comment by benelan** [comment]
I migrated to Lit's `isServer` constant from our previous `isBrowser()` function, which was introduced in #9897 to fix #9909. You can test `next` in the two repro samples from #9909 to verify that this change didn't cause a regression:

- [Gatsby repro sample](https://github.com/noahmulfinger/gatsby-calcite-components)
- [SvelteKit repro sample](https://github.com/feus4177/calcite-design-system-ssr-repro-9242)

**Comment by DitwanP** [comment]
🍠 Verified locally with samples provided.
